File: WiertarBot/commands/modules/alipaczka.py
from datetime import datetime
from typing import cast

import requests


class AliPaczka:
    __slots__ = ['tracking', 'number']

    def __init__(self, number: str) -> None:
        self.number = number
        self.tracking = self.get_tracking()

    def get_tracking(self) -> dict:
        p = requests.post(
            f'https://api.alipaczka.pl/track/{self.number}/',
            data={"uid": "2222", "ver": "22"}
        ).json()
        if 'error' not in p:
            for i in p['DataEntry']:
                i['time'] = datetime.fromtimestamp(int(i['time']))
        return cast(dict, p)

    def __str__(self):
        out = ''
        if 'error' in self.tracking:
            return self.tracking['error']

        out += f'Numer paczki: {self.number}\nDostarczona: '
        out += 'tak' if self.tracking['isDelivered'] else 'nie'
        for i in self.tracking['DataEntry']:
            out += f'\n{i["time"].strftime("%d/%m/%Y %H:%M")}: {i["status"]}'

        return out


if __name__ == "__main__":
    from sys import argv

    print(AliPaczka(argv[1]))

File: services/core/src/main/kotlin/pl/kvgx12/wiertarbot/commands/clients/external/AliPaczka.kt
package pl.kvgx12.wiertarbot.commands.clients.external

import kotlinx.datetime.Instant
import kotlinx.datetime.toLocalDateTime
import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import org.springframework.web.bind.annotation.PathVariable
import org.springframework.web.bind.annotation.RequestBody
import org.springframework.web.reactive.function.client.WebClientResponseException
import org.springframework.web.service.annotation.HttpExchange
import org.springframework.web.service.annotation.PostExchange
import pl.kvgx12.wiertarbot.commands.clients.external.AliPaczkaClient.TrackRequest
import pl.kvgx12.wiertarbot.commands.standard.plZone

@HttpExchange("https://api.alipaczka.pl")
interface AliPaczkaClient {
    @PostExchange("/track/{id}/")
    suspend fun track(
        @PathVariable id: String,
        @RequestBody request: TrackRequest,
    ): TrackResponse

    @Serializable
    data class TrackRequest(val uid: String, val ver: String)

    @Serializable
    data class TrackResponse(
        @SerialName("DataEntry")
        val entries: List<Entry>,
        val isDelivered: Boolean,
    ) {
        @Serializable
        data class Entry(val time: Long, val status: String)
    }
}

class AliPaczka(private val client: AliPaczkaClient) {
    suspend fun track(id: String): String {
        val response = try {
            client.track(id, TrackRequest("2222", "22"))
        } catch (_: WebClientResponseException.NotFound) {
            return "Nie znaleziono paczki"
        }

        return buildString {
            append(
                "Numer paczki: ",
                id,
                "\nDostarczono: ",
                if (response.isDelivered) "tak" else "nie",
            )
            response.entries.forEach {
                append(
                    "\n",
                    Instant.fromEpochSeconds(it.time)
                        .toLocalDateTime(plZone)
                        .run { "$dayOfMonth/$monthNumber/$year $hour:$minute" },
                    " - ",
                    it.status,
                )
            }
        }
    }
}
